### DISCUSSION TITLE
Help with regex matching syntax

### DISCUSSION THREAD

**Discussion by isaacdavidson1** [BurntSushi/ripgrep#1753] [Q&A] [answered]
Hello, I'm new to this crate entirely, and I'm trying to figure out how to use the grep crate to find matches of words like this (with `*` representing any number of characters):
 with a slice like this:
```rust
 const WORDLIST: &'static [u8] = b"\n word\n book\n sphinx\n someotherword\n"; ```

and then a search function like this:
```rust
fn search(pattern: &str, wordlist: &'static [u8]) -> Result<(), Box<Error>> {
    let matcher = RegexMatcher::new_line_matcher(pattern)?;
    let mut matches: Vec<(u64, String)> = vec![];
    Searcher::new().search_slice(&matcher, wordlist, UTF8(|lnum, line| {
        // We are guaranteed to find a match, so the unwrap is OK.
        let mymatch = matcher.find(line.as_bytes())?.unwrap();
        matches.push((lnum, line[mymatch].to_string()));
        Ok(true)
    }))?;
    println!("matches:{:?}", matches);
    Ok(())
}
```
And I want to be able to use a pattern like this `s*nx`  where the `*` would represent any number of characters, so one of the matches with search( `*s*nx`, WORDLIST) would be the word sphinx.
I'm not sure how to do this with regex syntax, and I took a look at https://docs.rs/regex/1.4.2/regex/#syntax but I am a huge beginner and couldn't figure it out. Any help would be greatly appreciated.

**Comment by BurntSushi** [comment, marked as answer]
Note: I edited your comment to use backticks instead of double quotes around your regexes. Without those, the `*` characters are usually interpreted as Markdown, which means it's rendered in a way you don't expect.

Note: I converted this to a discussion question since it seems more like a support question rather than a bug report or a feature request.

As for your question, it looks like you're asking for help with constructing the regex itself? And not the actual code? If so, you're pretty close. The `*` itself is a _unary operator_, meaning that it takes a single argument. In regex syntax, the `*` applies to the "thing" preceding it. In your case, `s*nx`, the `*` applies to the `s`, which modifies it to mean, "match the letter `s` zero or more times." Here are some other examples:

* `ab*c`, the `*` applies to `b` only.
* `(ab)*c`, the `*` applies to `ab`, and says, "match zero or more occurrences of of `ab`."

If you're used to globs, then this syntax is different. In globs, `*` is not an operator but a wildcard itself. The regex equivalent of it is more like, `.*`, which says, "match any character zero or more times." In regex syntax, the `.` is a wildcard meaning, "match any character."

So in your case, it _sounds_ like you want `s.*nx` instead of `s*nx`.

**Comment by isaacdavidson1** [comment]
Ah, ok. That makes sense. So potentially  I could also do something like
this: [a-z]* if I wanted to match one or more of any character between a
and z.

**Reply by BurntSushi** [reply]
Yes. I might suggest looking at a few regex tutorials to get your bearings. I don't have any particular one to recommend though unfortunately.

**Comment by isaacdavidson1** [comment]
Alrighty, thanks for the help.
